src/translator/service.cpp

```diff
@@ -28,10 +28,12 @@ loadVocabularies(marian::Ptr<marian::Options> options) {
 namespace marian {
 namespace bergamot {
 
-Service::Service(Ptr<Options> options, const void *model_memory)
+Service::Service(Ptr<Options> options, const void *model_memory, const size_t model_memory_size,
+                 const void *shortlist_memory, const size_t shortlist_memory_size)
     : requestId_(0), vocabs_(std::move(loadVocabularies(options))),
       text_processor_(vocabs_, options), batcher_(options),
-      numWorkers_(options->get<int>("cpu-threads")), model_memory_(model_memory)
+      numWorkers_(options->get<int>("cpu-threads")), model_memory_(model_memory), model_memory_size_(model_memory_size),
```

Comment:
I'd prefer the following different constructors instead of a bloated constructor: 

1, `Service(Options)` which purely loads from an options object, no bytearray business. This way (service-cli, service-cli-test-bytearray) can help figure if it's bytearray that's causing problems.
2, `Service(Options, ...)` incorporating all memory address + size passing instantiations. Do not default, if someone's calling this constructor, the compiler should complain that values are not passed. I'd like to bundle ... together so it's an all or nothing setting. The bytebuffers are passed for efficiency, doing just one doesn't make sense.
3. `Service(std::string config, ....)`: Do what saves you time with this one. Probably parseOptions which is already done, and pass through the memory address + size.

Use 3 in your test-app to test functionality covering both 2 and 3, and as this constructor is used in TranslationModel at the moment.

The default for data-members should be what makes bytearray loads no-op.

void* and size_t should be bound in a struct together and the struct given to Service not provided separately.

Reply:
Agree we should have a const void * and size_t blob struct.  It should also own the memory.

Reply:
I think @kpu has described what is desired from above more precisely already: https://github.com/browsermt/bergamot-translator/issues/63#issuecomment-804809630.

Reply:
I will put `MemoryGift` class into `definitions.h`.

Reply:
@jerinphilip it seems not possible to keep unified API boundaries (TranslationModel, AbstractTranslationModel) untouched. For the `service`'s constructors are changed (following what you said), I also need to adjust TranslationModel and AbstractTranslationModel.

Reply:
Hack the `string`-based constructor. It's only used by TranslationModel. Adding `MemoryGift` to WASM bound classes might have consequences (change wasm exports etc).

Reply:
@abhi-agg @motin @andrenatal Need some help here on getting your side to give us those binary blobs.

Reply:
@qianqianzhu You don't need to make any change to `AbstractTranslationModel` class as that class is gone (rebasing your PR on master will bring latest changes).

I can change the [constructor](https://github.com/browsermt/bergamot-translator/blob/main/src/TranslationModel.h#L36-L37) of `TranslationModel` class, push it to main branch and then you can rebase. However, the fastest way to unblock your PR would be if you just change this constructor to:

        TranslationModel(const std::string &config,
                           const void *model_memory = nullptr, size_t model_memory_size = 0,
                           const void *shortlist_memory = nullptr, size_t shortlist_memory_size = 0
                           )

You will not need to change anything elsewhere. Let me know if you are comfortable doing this change (and please feel free to ping me if it is not working for you 👍 )

Refactoring `const void *model_memory, size_t model_memory_size` with a `MemoryGift` struct or something can be done later once I generate JS bindings.

The bigger blocker of getting your PR approved is to fix the wasm CI failures and that has higher priority for me (compared to this small refactoring change) as I mentioned [here](https://github.com/browsermt/bergamot-translator/pull/69#issuecomment-806534147).

Reply:
Thanks! I hacked a way to solve the conflict of refactoring MemoryGift and TranslationModel. I will keep TranslationModel part untouched to not get in your way.